From 6c6e2accdfeb21e632702bf55cc458d291b50979 Mon Sep 17 00:00:00 2001
From: Alex Marginean <alexandru.marginean@nxp.com>
Date: Tue, 21 Nov 2017 16:46:15 +0200
Subject: [PATCH 084/706] enetc: add Rx C-TAG VLAN offload support

This is required for VLAN testing, current model in m166 sim only works
with VLAN offload enabled.

Signed-off-by: Alex Marginean <alexandru.marginean@nxp.com>
(cherry picked from commit 08cdeaf59e525fe3194cd4b01bde50a60948d5cb)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/net/ethernet/freescale/enetc/enetc.c    | 11 ++++++++++-
 drivers/net/ethernet/freescale/enetc/enetc_hw.h |  1 +
 2 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/freescale/enetc/enetc.c b/drivers/net/ethernet/freescale/enetc/enetc.c
index b63971a9206b..bbb59eb5c94a 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc.c
+++ b/drivers/net/ethernet/freescale/enetc/enetc.c
@@ -378,6 +378,13 @@ static void enetc_get_offloads(struct enetc_bdr *rx_ring,
 		skb->csum = csum_unfold((__force __sum16)~htons(inet_csum));
 		skb->ip_summed = CHECKSUM_COMPLETE;
 	}
+
+	/* copy VLAN to skb, if one is extracted, for now we assume it's a
+	 * standard TPID, but HW also supports custom values
+	 */
+	if (rxbd->r.flags & ENETC_RXBD_FLAG_VLAN)
+		__vlan_hwaccel_put_tag(skb, htons(ETH_P_8021Q),
+				       rxbd->r.vlan_opt);
 }
 
 #define ENETC_RXBD_BUNDLE 16 /* recommended # of BDs to update at once */
@@ -1318,7 +1325,9 @@ static void enetc_netdev_setup(struct enetc_si *si, struct net_device *ndev,
 	ndev->max_mtu = ENETC_MAX_MTU;
 
 	ndev->hw_features = NETIF_F_RXCSUM | NETIF_F_HW_CSUM;
-	ndev->features = ndev->hw_features | NETIF_F_HIGHDMA | NETIF_F_SG;
+	ndev->features = ndev->hw_features | NETIF_F_HIGHDMA | NETIF_F_SG |
+			 NETIF_F_HW_VLAN_CTAG_RX; /* < has to stay on for now */
+
 	ndev->priv_flags |= IFF_UNICAST_FLT;
 }
 
diff --git a/drivers/net/ethernet/freescale/enetc/enetc_hw.h b/drivers/net/ethernet/freescale/enetc/enetc_hw.h
index 699d476df6f3..03d1535bd5f3 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc_hw.h
+++ b/drivers/net/ethernet/freescale/enetc/enetc_hw.h
@@ -230,6 +230,7 @@ union enetc_rx_bd {
 #define ENETC_RXBD_LSTATUS_F	BIT(31)
 #define ENETC_RXBD_ERR_MASK	0xff
 #define ENETC_RXBD_LSTATUS(flags)	((flags) << 16)
+#define ENETC_RXBD_FLAG_VLAN	BIT(9)
 
 #define ENETC_MAC_ADDR_FILT_CNT	8 /* # of supported entries per port */
 #define ENETC_MAX_NUM_VFS	2
-- 
2.17.1

